src/prefect/orion/orchestration/core_policy.py

```diff
@@ -270,41 +274,96 @@ async def before_transition(
         proposed_state: Optional[states.State],
         context: FlowOrchestrationContext,
     ) -> None:
-        from prefect.orion.models import task_runs
-
         run_settings = context.run_settings
         run_count = context.run.run_count
+
         if run_settings.retries is None or run_count > run_settings.retries:
             return  # Retry count exceeded, allow transition to failed
 
         scheduled_start_time = pendulum.now("UTC").add(
             seconds=run_settings.retry_delay or 0
         )
 
-        failed_task_runs = await task_runs.read_task_runs(
-            context.session,
-            flow_run_filter=filters.FlowRunFilter(id={"any_": [context.run.id]}),
-            task_run_filter=filters.TaskRunFilter(state={"type": {"any_": ["FAILED"]}}),
-        )
-        for run in failed_task_runs:
-            await task_runs.set_task_run_state(
-                context.session,
-                run.id,
-                state=states.AwaitingRetry(scheduled_time=scheduled_start_time),
-                force=True,
-            )
-            # Reset the run count so that the task run retries still work correctly
-            run.run_count = 0
-
         # Generate a new state for the flow
         retry_state = states.AwaitingRetry(
             scheduled_time=scheduled_start_time,
             message=proposed_state.message,
             data=proposed_state.data,
         )
+
+        # support old-style flow run retries for older clients
+        # older flow retries require us to loop over failed tasks to update their state
+        # this is not required after API version 0.8.3
+        api_version = context.parameters.get("api-version", None)
```

Comment:
this implies to me that the new logic is incompatible with old clients but abstractly I don't see why that should be the case - could you help me understand that?

Reply:
It's really minor actually—we require a small client change to handle subflows (hence the new integration test, they all would have passed before its addition). Because we want the subflow to be rerun (as per the existing tests that define flow retry behavior) we need to handle it here: https://github.com/PrefectHQ/prefect/blob/flow-restarts/src/prefect/engine.py#L423-L425

Reply:
I should add that if we don't enforce subflow retry behavior to be preserved exactly we can remove this compatibility code, flows with just tasks will work just as well on old clients

Reply:
I think it's OK to keep around for a bit, thanks for the clarification!